core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/util/MetricUtils.java

```diff
@@ -54,6 +56,19 @@ private MetricUtils() {
         return Optional.ofNullable(clientConfig.option(METRIC_PUBLISHER));
     }
 
+    /**
+     * Resolve the correct metric publisher to use. The publisher set on the request always takes precedence.
+     *
+     * @param clientConfig The client configuration.
+     * @param requestConfig The request override configuration.
+     * @return The metric publisher to use.
+     */
+    public static Optional<MetricPublisher> resolvePublisher(SdkClientConfiguration clientConfig,
```

Comment:
I created this overload because for some operations, the input request is modified (eg: applySignerOverride), so the request variable is not final. As a result, we can't use the input request parameter in the `whenComplete` block because variables used in lambda should be final or effectively final

Reply:
Not a fan of using `Optional` as a method parameter. Can we change to so it takes an override config, and change the operation code:

```java
MetricUtils.resolvePublisher(clientConfig, overrideConfig.orElse(null));
```
```

Reply:
In most cases, I'd agree on the use of optional as a method parameter, but I can't find a better way to handle it in this case.

`MetricUtils.resolvePublisher(clientConfig, overrideConfig.orElse(null));` seems a bit code smell to me. 

Ideally, we should create another overload which doesn't take a requestOverrideConfig, but it seems overkill here.
```java
MetricUtils.resolvePublisher(SdkClientConfiguration clientConfig, AwsRequestOverrideConfiguration overrideConfig)
MetricUtils.resolvePublisher(SdkClientConfiguration clientConfig)
```